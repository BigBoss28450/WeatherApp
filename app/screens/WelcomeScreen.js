import React, { useState } from 'react';
import { Alert, FlatList, Modal, Text, View } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import CityWeatherButton from '../components/CityWeatherButton';
import Screen from '../components/Screen';
import HeaderBar from '../components/headerComponents/HeaderBar';
import weatherApiCalls from '../api/weatherApiCalls';
import ModalContainer from '../components/ModalContainer';
import MainWeatherComponent from '../components/MainWeatherComponent';
import TextButton from '../components/basicComponents/TextButton';
import CustomTextInput from '../components/basicComponents/CustomTextInput';
import HeaderButton from '../components/headerComponents/HeaderButton';
import { getWeather } from '../redux/actions/weather';
import colors from '../config/colors';

export default function WelcomeScreen({ navigation }) {
  // states
  const [modalVisible, setModalVisible] = useState(false);
  const [tempWeather, setTempWeather] = useState(null);
  const [noResults, setNoResults] = useState(false);

  // variables
  const weatherList = useSelector((state) => state.weather.weatherList);
  const dispatch = useDispatch();

  // functions
  const searchCityWeather = async (city) => {
    try {
      const result = await weatherApiCalls.getCityWeather(city);
      if (result.cod === 200) {
        setNoResults(false);
        setTempWeather(result);
      }
      if (result.cod === '404') {
        setTempWeather(null);
        setNoResults(true);
      }
      return null;
    } catch (error) {
      Alert.alert('Â¡Error!', 'No estas conectado a internet');
      return null;
    }
  };

  const addToWeatherList = (data) => {
    dispatch(getWeather(data));
    setTempWeather(null);
    setModalVisible(false);
    return null;
  };

  return (
    <Screen>
      <View style={{ flex: 1 }}>
        <HeaderBar>
          <HeaderButton icon="magnify" onPress={() => setModalVisible(true)} />
        </HeaderBar>
        <View style={{ flex: 1 }}>
          <FlatList
            contentContainerStyle={{
              flexGrow: 1,
              justifyContent: 'center',
              paddingVertical: 20,
            }}
            numColumns={2}
            data={weatherList}
            keyExtractor={(item, index) => index.toString()}
            renderItem={({ item }) => (
              <CityWeatherButton
                city={item.name}
                temp={item.main.temp}
                onPress={() =>
                  navigation.navigate('Details', { city: item.name })
                }
              />
            )}
          />
        </View>
      </View>
      <Modal visible={modalVisible} transparent animationType="slide">
        <ModalContainer>
          <HeaderBar>
            <CustomTextInput
              onSubmitEditing={(string) => searchCityWeather(string)}
            />
            <HeaderButton
              icon="close"
              onPress={() => {
                setTempWeather(null);
                setNoResults(false);
                setModalVisible(false);
                return null;
              }}
            />
          </HeaderBar>
          {tempWeather && (
            <>
              <HeaderBar>
                <TextButton
                  title="Agregar a la lista"
                  onPress={() => addToWeatherList(tempWeather)}
                />
              </HeaderBar>
              <MainWeatherComponent
                city={tempWeather.name}
                description={tempWeather.weather[0].description}
                temp={tempWeather.main.temp}
                max={tempWeather.main.temp_max}
                min={tempWeather.main.temp_min}
              />
            </>
          )}
          {noResults && (
            <Text style={{ color: colors.white, textAlign: 'center' }}>
              No se encontraron resutados.
            </Text>
          )}
        </ModalContainer>
      </Modal>
    </Screen>
  );
}

WelcomeScreen.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,
  }).isRequired,
};
