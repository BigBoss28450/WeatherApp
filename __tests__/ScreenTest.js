import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import WelcomeScreen from '../app/screens/WelcomeScreen';
import configureStore from '../app/redux/store';

jest.useFakeTimers();

test('Welcome Screen renders correctly', () => {
  const store = configureStore();
  const mainScreen = renderer
    .create(
      <Provider store={store}>
        <WelcomeScreen />
      </Provider>
    )
    .toJSON();
  expect(mainScreen).toMatchSnapshot();
});
