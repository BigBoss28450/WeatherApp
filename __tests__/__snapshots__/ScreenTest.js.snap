// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome Screen renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.3)",
        "flex": 1,
      }
    }
  >
    <RNCSafeAreaView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignSelf": "flex-end",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "maxHeight": 60,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "height": "100%",
                  "justifyContent": "center",
                  "width": 50,
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <RCTScrollView
            contentContainerStyle={
              Object {
                "flexGrow": 1,
                "justifyContent": "center",
                "paddingVertical": 20,
              }
            }
            data={Array []}
            disableVirtualization={false}
            getItem={[Function]}
            getItemCount={[Function]}
            horizontal={false}
            initialNumToRender={10}
            keyExtractor={[Function]}
            maxToRenderPerBatch={10}
            onContentSizeChange={[Function]}
            onEndReachedThreshold={2}
            onLayout={[Function]}
            onMomentumScrollEnd={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            removeClippedSubviews={false}
            renderItem={[Function]}
            scrollEventThrottle={50}
            stickyHeaderIndices={Array []}
            updateCellsBatchingPeriod={50}
            viewabilityConfigCallbackPairs={Array []}
            windowSize={21}
          >
            <View />
          </RCTScrollView>
        </View>
      </View>
      <Modal
        animationType="slide"
        hardwareAccelerated={false}
        transparent={true}
        visible={false}
      >
        <View
          style={
            Object {
              "flex": 1,
              "justifyContent": "flex-end",
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#000000",
                "flex": 1,
                "maxHeight": 800,
              }
            }
          >
            <View
              style={
                Object {
                  "alignSelf": "flex-end",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "maxHeight": 60,
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "paddingLeft": 20,
                    "paddingVertical": 10,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 20,
                      "flex": 1,
                      "justifyContent": "center",
                      "paddingHorizontal": 20,
                    }
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    onChangeText={[Function]}
                    onSubmitEditing={[Function]}
                    placeholder="Aa"
                    rejectResponderTermination={true}
                    returnKeyLabel="Buscar"
                    returnKeyType="search"
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                    underlineColorAndroid="transparent"
                  />
                </View>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "height": "100%",
                      "justifyContent": "center",
                      "width": 50,
                    }
                  }
                >
                  <Text />
                </View>
              </View>
            </View>
          </View>
        </View>
      </Modal>
    </RNCSafeAreaView>
  </View>
</View>
`;
